### ISSUE TITLE
spec: non-equivalence of instantiated, generic functions which are instantiated to be equivalent

### ISSUE THREAD

**Issue by sammy-hughes** [golang/go#64664]
### Go version

go version go1.21.5 darwin/arm64

### What operating system and processor architecture are you using (`go env`)?

```shell
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/*redacted*/Library/Caches/go-build'
GOENV='/Users/*redacted*/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/*redacted*/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/*redacted*/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.21.5/libexec'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.21.5/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.21.5'
GCCGO='gccgo'
AR='ar'
CC='cc'
CXX='c++'
CGO_ENABLED='1'
GOMOD='/Users/*redacted*/githubrepos/*redacted*/go.mod'
GOWORK='/Users/*redacted*/githubrepos/*redacted*/go.work'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/77/q6n8zv350gsc0f0b221l5xl00000gp/T/go-build1394031279=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

### What did you do?

https://go.dev/play/p/OZg7DTVu-bn

If it would help, I could copy-paste the contents of that here. The summary is that I was writing some generic wrappers around some helper functions which used a generic function type having a different structure, but which is by definition compatible. There's two smaller complaints which are, I think discussed elsewhere, so the only complaint here is that the two instantiated template function types should be fully domain-equivalent, but require a different, explicit cast as a non-generic type for assignment one to another.

### What did you expect to see?

With the following example, `fnA = fnB` or `fnB = fnA` should be resolved, given that `fnA = (func(map[string]any, map[string]any)bool)(fnB)` and also `fnB = (func(map[string]any, map[string]any)bool)(fnA)` are both correctly resolved. 
```Go
// btree.LessFunc[T any](T, T) bool
var fnA btree.LessFunc[map[string]any]

// LessFunc[K comparable, M map[K]any](M, M) bool
var fnB LessFunc[string, map[string]any]
```

### What did you see instead?

With the following example, `fnA = fnB` and `fnB = fnA` are not resolved as compatible and can only be cross-assigned with an explicit conversion using a concrete function signature which by definition matches both instantiated template types.
```Go
// btree.LessFunc[T any](T, T) bool
var fnA btree.LessFunc[map[string]any]

// LessFunc[K comparable, M map[K]any](M, M) bool
var fnB LessFunc[string, map[string]any]
```

**Comment by prattmic** [comment]
I've tentatively marked this a "spec", but it isn't clear to me if this is a bug in the spec, a bug in the compiler, or working as intended (and thus should be a proposal).

cc @griesemer @findleyr @ianlancetaylor

**Comment by griesemer** [comment]
Glancing at this superficially, it looks like `fnA` and `fnB` have two named (instantiated) types that are differently named, so they are different, and a conversion is needed.

Maybe there's more here, so can you please add a short, self-contained (no reliance on external libraries), minimal piece of code (10-20 lines, playground example) that illustrates the problem directly, so that we don't have to distill a 150+ line example down to its essence?

**Comment by findleyr** [comment]
This is working as intended. LessFunc and btree.LessFunc are differently named, as @griesemer suggested. IIUC, the following example is equivalent, and doesn't involve generics:
https://go.dev/play/p/gGs-gk--0Po

**Comment by sammy-hughes** [comment]
Going into the holidays, I totally forgot about this. I'm very confused about how this was handled, though apologies for non-responsiveness. I believe there is clearly further discussion here, and would like to know if this issue can be reopened. If not, I can re-create the issue anew. 

Meanwhile, in response to the above comments

1. This issue was renamed as a "spec" issue. I'm unclear on why or what that means. I mentioned generics instantiation as the behavior I expected was that differently-templated but fully instantiated generic function signatures would be assignment compatible and not just conversion compatible.
2. It is complained that the snippet depends on external libraries. The btree library referenced is a google library, though yes, it's still an "external library", sure. I care about this issue for the sake of working with libraries implemented using generics, and very specifically THIS library.
3. It is complained that the snippet is long. I am demonstrating both the issue, while also demonstrating how I currently have to work around the issue. This could instead be 3 different snippets, sure. I can do that. See below.
4. It is implied that my confusion was over how I created a function type which is similarly named to an imported function type name. I didn't expect that to be confusing, but I repeat, my complaint is that something like `func[T comparable](map[T]any)bool` isn't directly assignment compatible with something like `func[T comparable, M map[T]any](M)bool`, when specifically instantiated such that they are both compatible with the same simple function signature. The similar naming is to highlight the intended compatibility, not an expectation that it magically makes it compatible?

The snippet I included originally was intended to demonstrate that the Go compiler fails to resolve two different generic template function types as being entirely compatible, when instantiated with the same context. There's some additional code included to show something about why this matters. I have reformulated it to appear in 3 separate go-playground snippets. Example 1 doesn't compile, and examples 2 and 3 show why example 1 should be fine.
- Example 1: https://go.dev/play/p/Nh2qLpvvMC3
- Example 2: https://go.dev/play/p/G5VUCDGJ2Q5
- Example 3: https://go.dev/play/p/wwKiNZgpoS_6
